### PR TITLE
QOI support

### PR DESCRIPTION
**Pull request by ehmry**
https://qoiformat.org/

---

Not actually lossless when converting the test reference files. There is some sort of off-by-one shift during `toPremultipliedAlpha` and `toStraightAlpha`. I see alpha artifacts on the edges of [testcard_rgba.png](https://raw.githubusercontent.com/treeform/pixie/ae2bef5606a275eacdc6cf27a402c1baf47a2b14/tests/fileformats/qoi/testcard_rgba.png) that are somehow missing from the Pixie roundtrip encoding.

### CODE REVIEW

**Comment by treeform** [comment]
The toPremultipliedAlpha <> toStraightAlpha are not lossless. We run into this problem as well. In the png decoder we have a "raw" `Png` type that is lossless though. You might create a Qoi type that is losses and do most tests wit that. Then have the Image to Qoi conversation that does do the premultiplied alpha is another layer.

**Comment by ehmry** [comment]
Done.

I still don't understand the what the "sRBG with linear alpha" or "all channels linear" bit flag implies so I'm not doing anything with it.

File: src/pixie/fileformats/qoi.nim [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  Header = tuple
+    width, height: uint32
+    channels, colorspace: uint8
+  Colorspace* = enum sRBG = 0, linear = 1
+  Index = array[indexLen, ColorRGBX]
+
+func hash(p: ColorRGBX): int =
+  (p.r.int * 3 + p.g.int * 5 + p.b.int * 7 + p.a.int * 11) mod indexLen
+
+proc decodeQoiRaw*(data: string): (Image, Colorspace) {.raises: [PixieError].} =
```
Comment (guzba) [inline src/pixie/fileformats/qoi.nim:27]: Please do not use an `Image` to hold straight alpha data. It does not ever hold that. If you want to hold straight alpha values, please make a type for that similar to `Png` in the png.nim file.

**Comment by guzba** [comment]
I've not personally used the Quite OK image format, but I have heard of it and supporting it seems cool.

File: src/pixie/fileformats/qoi.nim (hunk 1 above)
Comment (ehmry) [reply src/pixie/fileformats/qoi.nim]: Good idea.

**Comment by ehmry** [comment]
Pushed a new version with a `Qoi` data type.

I was thinking we have enough image formats but after these NSO image parser exploits I can appreciate something that is designed to be simple.

**Comment by treeform** [comment]
Thank you for your PR! This is great. 

If you want to add any other image formats let us know.

**Comment by digitalcraftsman** [comment]
This great addition may also be mentioned in the [README](https://github.com/treeform/pixie#image-file-formats)

**Pull request merged**